### PR TITLE
Introduce cargo makefiles

### LINKED ISSUE #299: Move CI over to cargo-make
ICU4X uses cargo-make for CI and general tasks and it's quite convenient.

Currently working on diplomat requires running shell scripts from various folders and it's not super fine grained, it's annoying.

Running the main build via `cargo build` seems fine, but we should have tasks for:

 - `gen-{example, feature}-{c, cpp, ...}`. All of these should also perhaps regenerate in a separate folder and only copy over when successful.
 - `gen-{example, feature}`
 - `gen-{c, cpp, ...}`
 - `self-contained-c2` (and more as we do more tests like this)
 - `test-feature-cpp`, `test-example-cpp`, `test-feature-js`, etc.

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes #299

<s>This doesn't yet move everything over, but it improves the situation around regeneration. I might move more stuff over</s>

CI is now effectively running on makefiles. Huzzah!

Note that the generated files change a little bit since there were old checked in files that should have been deleted. With the new code this will no longer happen.

### CODE REVIEW

**Review by sffc — APPROVED** [review]

File: Makefile.toml [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+# Generic task for generating code
+# Usage: cargo make gen-generic feature_tests <backend name> <output folder name> <optional docs folder> <optional additional args>
+[tasks.gen-generic]
+category = "Plumbing"
+dependencies = ["build-tool"]
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+
+path = array_get ${@} 0
```
Comment (sffc) [inline Makefile.toml:201]: Praise: Glad you got Makefile arguments working

File: Makefile.toml [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+[tasks.test-c2-example-self-contained]
+category = "Tests"
+dependencies = ["build-example"]
+script_runner = "@duckscript"
+script = '''
+exit_on_error true
+cd example/c2/include
+files_list = glob_array "./*.h"
+for file in ${files_list}
+    exec gcc ${file}
```
Comment (sffc) [inline Makefile.toml:173]: Thought: Should `gcc` be pulled out?

File: Makefile.toml (hunk 1 above)
Comment (Manishearth) [reply Makefile.toml:201]: it's an absolute mess lol, but it works

Comment (Manishearth) [reply Makefile.toml:201]: https://github.com/rust-diplomat/diplomat/pull/304

**Pull request merged**